pkg/expression/builtin_time.go

```diff
@@ -280,7 +280,8 @@ func (b *builtinDateSig) evalTime(ctx sessionctx.Context, row chunk.Row) (types.
 		return types.ZeroTime, true, handleInvalidTimeError(b.ctx, err)
 	}
 
-	if expr.IsZero() && b.ctx.GetSessionVars().SQLMode.HasNoZeroDateMode() {
+	if expr.IsZero() && b.ctx.GetSessionVars().SQLMode.HasNoZeroDateMode() ||
```

Comment:
I think it's not an issue about `date` function. It's about the conversion from `year` to `datetime`. For example, the following SQL will still have different behavior (with MySQL and TiKV) with this PR:

```
select cast(cast(1999 as year) as datetime);
```

It's expected to give a warning, but TiDB gives `1999-00-00 00:00:00`. I think it'd be better to fix `builtinCastIntAsTimeSig` rather than `date` function.

Reply:
Oops. It seems that it's also incorrect to just report warning when converting year to datetime, because it doesn't match the expected behavior mentioned in https://github.com/pingcap/tidb/pull/20233 :facepalm: . Conversion has different behavior in comparison...

Reply:
[try_year_cmp_func](https://github.com/mysql/mysql-server/blob/87307d4ddd88405117e3f1e51323836d57ab1f57/sql/item_cmpfunc.cc#L1537).

Reply:
1. change `year` compare by convert it to `ulonglong` as mysql dose.
2. make cast `year` to `datetime` and `date` always invalid
How about hanle it like above?@YangKeao

Reply:
1. change year compare by convert it to ulonglong as mysql does.

IIUC, [try_year_cmp_func](https://github.com/mysql/mysql-server/blob/87307d4ddd88405117e3f1e51323836d57ab1f57/sql/item_cmpfunc.cc#L1537) shows they are converted to `datetime` but not `ulonglong`. It's just a special behavior. The `get_year_value` function wraps a `year` into a `datetime` (which is represented as `longlong` inside). The behavior of converting a `year` to `unsigned` (in SQL expression) is using the year directly. For example, `cast(cast(2022 as year) as unsigned)` will give you `2022`. So, return `types.ETDatetime` is expected in `getBaseCmpType` for this situation.